pkg/api/v2/join.go

```diff
@@ -0,0 +1,209 @@
+	// CertificateAuthority is the root CertificateAuthority certificate for the Kubernetes cluster.
+	CertificateAuthority string `json:"ca"`
+	// CallbackToken is a callback token used to authenticate requests with the cluster agent on the joining node.
+	CallbackToken string `json:"callback_token"`
+	// APIServerPort is the port where the kube-apiserver is listening.
+	APIServerPort string `json:"apiport"`
+	// APIServerAuthorizationMode is the AuthorizationMode used by kube-apiserver.
+	APIServerAuthorizationMode string `json:"api_authz_mode"`
+	// KubeletArgs is a string with arguments for the kubelet service on the joining node.
+	KubeletArgs string `json:"kubelet_args"`
+	// HostNameOverride is the host name the joining node will be known as in the MicroK8s cluster.
+	HostNameOverride string `json:"hostname_override"`
+	// DqliteVoterNodes is a list of known dqlite voter nodes. Each voter is identified as "$IP_ADDRESS:$PORT".
+	// This is not included in the response when joining worker-only nodes.
+	DqliteVoterNodes []string `json:"voters,omitempty"`
+	// ServiceAccountKey is the private key used for signing ServiceAccount tokens.
+	// This is not included in the response when joining worker-only nodes.
+	ServiceAccountKey string `json:"service_account_key"`
+	// AdminToken is a static token used to authenticate in the MicroK8s cluster as "admin".
+	// This is not included in the response when joining worker-only nodes.
+	AdminToken string `json:"admin_token,omitempty"`
+	// CertificateAuthorityKey is the private key of the Certificate Authority.
+	// Note that this is defined as *string, since the Python code expects this to be explicitly None/nil/null for worker only nodes.
+	// This is not included in the response when joining worker-only nodes.
+	CertificateAuthorityKey *string `json:"ca_key"`
+	// DqliteClusterCertificate is the certificate for connecting to the Dqlite cluster.
+	// This is not included in the response when joining worker-only nodes.
+	DqliteClusterCertificate string `json:"cluster_cert,omitempty"`
+	// DqliteClusterKey is the key for connecting to the Dqlite cluster.
+	// This is not included in the response when joining worker-only nodes.
+	DqliteClusterKey string `json:"cluster_key,omitempty"`
+	// ControlPlaneNodes is a list of known control plane nodes running kube-apiserver.
+	// This is only included in the response when joining worker-only nodes.
+	ControlPlaneNodes []string `json:"control_plane_nodes"`
+}
+
+// Join implements "POST v2/join".
+func (a *API) Join(ctx context.Context, req JoinRequest) (*JoinResponse, error) {
+	if !util.IsValidClusterToken(req.ClusterToken) {
+		return nil, fmt.Errorf("invalid cluster token")
+	}
+	if err := util.RemoveClusterToken(req.ClusterToken); err != nil {
+		return nil, fmt.Errorf("failed to remove cluster token: %w", err)
+	}
+	if !util.HasDqliteLock() {
+		return nil, fmt.Errorf("failed to join the cluster: this is not an HA MicroK8s cluster")
```

Comment:
Lets keep the same error message and the same http error code as we have in the python implementation.